File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    n = len(friends)
    
    arr = [[0] * n for _ in range(n)]
    gift_score = [0]*n
    
    
    for i in gifts:
        idx = i.index(' ')
        sender = i[:idx]
        receiver = i[idx+1:]
        arr[friends.index(sender)][friends.index(receiver)] += 1
        arr[friends.index(receiver)][friends.index(sender)] -= 1
        gift_score[friends.index(sender)] += 1
        gift_score[friends.index(receiver)] -= 1
    result = 0
    for idx, friend in enumerate(friends):
        score = gift_score[idx]
        gift = arr[idx]
        cur = sum([1 for i in gift if i > 0])
        cur += sum([1 for index,i in enumerate(gift) if i == 0 and score > gift_score[index]])
        result = max(result, cur)
    return result
        
        

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.*;

class Solution {
  public int solution(String[] friends, String[] gifts) {
      int n = friends.length;

      Map<String, Integer> friendIndex = new HashMap<>();
      for (int i = 0; i < n; i++) {
          friendIndex.put(friends[i], i);
      }

      int[][] arr = new int[n][n];
      int[] giftScore = new int[n];

      for (String gift : gifts) {
          String[] parts = gift.split(" ");
          String sender = parts[0];
          String receiver = parts[1];

          int senderIdx = friendIndex.get(sender);
          int receiverIdx = friendIndex.get(receiver);

          arr[senderIdx][receiverIdx] += 1;
          arr[receiverIdx][senderIdx] -= 1;
          giftScore[senderIdx] += 1;
          giftScore[receiverIdx] -= 1;
      }

      int result = 0;
      for (int idx = 0; idx < n; idx++) {
          int score = giftScore[idx];
          int[] gift = arr[idx];

          int cur = 0;
          for (int i = 0; i < n; i++) {
              if (gift[i] > 0) {
                  cur++;
              } else if (gift[i] == 0 && score > giftScore[i]) {
                  cur++;
              }
          }

          result = Math.max(result, cur);
      }

      return result;
  }
}